doc/sg/guidance_for_volatile.md

```diff
@@ -0,0 +1,153 @@
+---
+title: "Guidance for `volatile` in OpenTitan Silicon Creator Code"
+---
+
+## TL;DR
+
+Do not use `volatile` in production, i.e. non-test, silicon creator code unless you are implementing a library explicitly for this purpose like `sec_mmio`, `abs_mmio`, or `hardened`.
+Specifically, do not use `volatile` for
+
+*   Accessing memory-mapped registers,
+    *   _Instead, use `sec_mmio` or `abs_mmio`_.
+*   Hardening purposes.
+    *   _Instead, use the primitives in `hardened.h`._
+*   Variables initialized outside the lifetime of the program, e.g. a pointer to some data in flash or retention SRAM, or
+    *   _Instead, declare as non-`const `and non-`volatile`. Use a getter to enforce `const`-ness if needed._
+*   Pointers created out of thin air.
+    *   _Compiler cannot optimize the first access (assuming that the pointer doesn't point to something already known by the compiler)._
+
+Do use `volatile` if a variable is shared between an interrupt/exception handler and the rest of the on-target test program to ensure correctness.
+*   _Since silicon creator code does not use handlers in the conventional sense, this essentially limits `volatile` usage to on-target tests._
+
+When in doubt, please do not hesitate to reach out by creating a GitHub issue (preferably with the "Type:Question" label).
+
+## Introduction
+
+The goal of this document is to provide guidance for using `volatile` in OpenTitan Silicon Creator code, i.e. `rom`, `rom_ext`, related tests, and examples.
+There are several reasons for this guidance:
+
+*   `volatile` is contagious and, similar to `const`-correctness, `volatile`-correctness can be hard to achieve and maintain. Once a variable is declared as `volatile`, the `volatile` keyword must appear throughout the call stack.
+*   `volatile` variables cannot be passed as arguments to standard functions such as `memcpy`, `memset`, etc. Casting `volatile`-ness away in such cases results in undefined behavior. This is an easy-to-make but hard-to-catch mistake.
```

Comment:
Reading this again, it might be better to clarify here that these points apply to pointers and references only. E.g. something like

```suggestion
*   `volatile` is contagious and, similar to `const`-correctness, `volatile`-correctness can be hard to achieve and maintain. Once a variable is declared as `volatile`, the `volatile` keyword must appear throughout the call stack in all other variables that will reference the same object.
*   `volatile` pointers cannot be passed as arguments to standard functions such as `memcpy`, `memset`, etc. Casting `volatile`-ness away in such cases results in undefined behavior. This is an easy-to-make but hard-to-catch mistake.
```

Otherwise it would seem to imply that something like this is invalid:

```
volatile int x;
void foo(int);
void bar(void) {
    foo(x);
}
```

Reply:
Good point, I agree with your suggestions, they are more precise.